src/rdkafka_metadata_cache.c

```diff
@@ -427,24 +489,38 @@ void rd_kafka_metadata_cache_topic_update(
     rd_bool_t include_racks,
     rd_kafka_metadata_broker_internal_t *brokers,
     size_t broker_cnt) {
-        rd_ts_t now        = rd_clock();
+        struct rd_kafka_metadata_cache_entry *rkmce = NULL;
+        rd_ts_t now                                 = rd_clock();
         rd_ts_t ts_expires = now + (rk->rk_conf.metadata_max_age_ms * 1000);
         int changed        = 1;
+        if (unlikely(!mdt->topic)) {
+                rkmce =
+                    rd_kafka_metadata_cache_find_by_id(rk, mdit->topic_id, 1);
+                if (!rkmce)
+                        return;
+        }
 
-        /* Cache unknown topics for a short while (100ms) to allow the cgrp
-         * logic to find negative cache hits. */
-        if (mdt->err == RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_OR_PART)
-                ts_expires = RD_MIN(ts_expires, now + (100 * 1000));
-
-        if (!mdt->err ||
-            mdt->err == RD_KAFKA_RESP_ERR_TOPIC_AUTHORIZATION_FAILED ||
-            mdt->err == RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_OR_PART)
-                rd_kafka_metadata_cache_insert(rk, mdt, mdit, now, ts_expires,
-                                               include_racks, brokers,
-                                               broker_cnt);
-        else
-                changed =
-                    rd_kafka_metadata_cache_delete_by_name(rk, mdt->topic);
+        if (!mdt->topic) {
+                /* Cache entry found but no topic name:
+                 * delete it. */
+                changed = rd_kafka_metadata_cache_delete_by_topic_id(
+                    rk, mdit->topic_id);
+        } else {
+                /* Cache unknown topics for a short while (100ms) to allow the
+                 * cgrp logic to find negative cache hits. */
+                if (mdt->err == RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_OR_PART)
+                        ts_expires = RD_MIN(ts_expires, now + (100 * 1000));
+
+                if (!mdt->err ||
+                    mdt->err == RD_KAFKA_RESP_ERR_TOPIC_AUTHORIZATION_FAILED ||
+                    mdt->err == RD_KAFKA_RESP_ERR_UNKNOWN_TOPIC_OR_PART)
+                        rd_kafka_metadata_cache_insert(
+                            rk, mdt, mdit, now, ts_expires, include_racks,
+                            brokers, broker_cnt);
+                else
+                        changed = rd_kafka_metadata_cache_delete_by_name(
+                            rk, mdt->topic);
```

Comment:
Shall we add `delete_by_topic_id` as well here? Just give it a thought if it is required.

Reply:
No, because it finds by name and then calls `rd_kafka_metadata_cache_delete` that deletes from both